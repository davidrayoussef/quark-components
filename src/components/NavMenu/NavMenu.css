.menuTitle {
  margin: 0 0 5px;
  font-size: 1rem;
}

.nav {
  display: flex;
  flex-direction: column;
}

.nav a {
  text-decoration: none;
  margin-bottom: 5px;
}
