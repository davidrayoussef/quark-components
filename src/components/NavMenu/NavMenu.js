import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import style from './NavMenu.css';
import { computeHoverColor, computeActiveColor } from '../../utils/computeHoverColor';

class NavMenu extends Component {
  static defaultProps = {
    linkColor: 'rgb(52, 88, 226)',
    linkType: 'anchorLink',
    linkDisabled: false
  };

  static propTypes = {
    data: PropTypes.array.isRequired,
    linkColor: PropTypes.string,
    linkType: PropTypes.string,
    linkDisabled: PropTypes.bool,
    menuTitle: PropTypes.string
  };

  handleClick = (e) => {
    if (this.props.linkDisabled) e.preventDefault();
  };

  render() {
    const { data, linkColor, linkType, menuTitle } = this.props;
    const renderListItems = () => {
      if (linkType === 'routerLink') {
        return data.map(({name}) => (
          <NavLink
            key={name}
            to={'/' + name.toLowerCase()}
            style={{ color: linkColor }}
            activeStyle={{ borderColor: linkColor, background: computeActiveColor(linkColor) }}
            onClick={this.handleClick}
          >
            {name}
          </NavLink>
        ));
      }
      return data.map(({ name }) => (
        <a
          key={name}
          href={'/' + name.toLowerCase()}
          style={{ color: linkColor }}
          onClick={this.handleClick}
        >
          {name}
        </a>
      ));
    };

    return (
      <Fragment>
        <h2 className={style.menuTitle}>{menuTitle.toUpperCase()}</h2>
        <nav className={style.nav}>
          { renderListItems() }
        </nav>
        <style>{ computeHoverColor(linkColor) }</style>
      </Fragment>
    );
  }
}

export default NavMenu;
