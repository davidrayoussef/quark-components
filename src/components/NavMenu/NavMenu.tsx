import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';

import { NavMenuProps } from '../../shared';
import {
  computeHoverColor,
  computeActiveColor,
  isMobile
} from '../../utilities';

import style from './NavMenu.css';

export class NavMenu extends Component<NavMenuProps> {
  static defaultProps = {
    linkColor: 'rgb(52, 88, 226)',
    linkType: 'anchorLink',
    linkDisabled: false
  };

  handleLinkClick = (event: React.MouseEvent): void => {
    const { linkDisabled, handleMobileLinkClick } = this.props;
    if (linkDisabled) event.preventDefault();
    if (handleMobileLinkClick) handleMobileLinkClick();
  };

  render(): JSX.Element {
    const { data, linkColor, linkType, menuTitle, linkDisabled } = this.props;
    const renderListItems = () => {
      if (linkType === 'routerLink') {
        return data.map(({ name }) => (
          <NavLink
            exact
            to={'/' + name.toLowerCase()}
            key={name}
            style={{ color: linkColor }}
            activeStyle={{
              borderColor: linkColor,
              background: computeActiveColor(linkColor)
            }}
            onClick={this.handleLinkClick}
          >
            {name}
          </NavLink>
        ));
      }
      return data.map(({ name }) => (
        <a
          key={name}
          href={linkDisabled ? '#' : '/' + name.toLowerCase()}
          style={{ color: linkColor }}
          onClick={this.handleLinkClick}
        >
          {name}
        </a>
      ));
    };
    return (
      <aside
        className={style.sidebar}
        style={{
          ...(this.props.mobileStyles ?? {}),
          display:
            !isMobile() ||
            !this.props.hasOwnProperty('isOpen') ||
            (isMobile() && this.props.isOpen)
              ? 'block'
              : 'none'
        }}
      >
        {menuTitle && (
          <h2 className={style.menuTitle}>{menuTitle.toUpperCase()}</h2>
        )}
        <nav className={style.nav}>{renderListItems()}</nav>
        <style>{computeHoverColor(linkColor)}</style>
      </aside>
    );
  }
}
