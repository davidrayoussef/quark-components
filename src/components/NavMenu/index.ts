export { NavMenu } from './NavMenu';
export { withMobileMenu } from './withMobileMenu';
