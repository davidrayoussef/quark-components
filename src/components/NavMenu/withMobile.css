.menu-icon {
  display: none;
  cursor: pointer;
  position: fixed;
  top: 12px;
  right: 0;
}

.menu-icon path {
  transition: .15s ease-in-out;
}

.menu-icon .close.line1 {
  transform: rotate(45deg) translateX(6px) translateY(-6px);
}

.menu-icon .close.line2 {
  transform: rotate(45deg) translateX(6px) translateY(-6px);
}

.menu-icon .close.line2 {
  transform: rotate(45deg) translateX(6px) translateY(-6px);
}

@media (max-width: 800px) {
  .menu-icon {
    display: block;
  }
}
