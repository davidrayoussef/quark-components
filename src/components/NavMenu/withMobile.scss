.menu-icon {
  display: none;
  cursor: pointer;
  position: fixed;
  top: 12px;
  right: 0;
  z-index: 2;
  & path {
    transition: 0.15s ease-in-out;
  }
  &.close {
    & path:nth-child(1) {
      transform: rotate(45deg) translateX(6px) translateY(-6px);
    }
    & path:nth-child(2) {
      opacity: 0;
      visibility: hidden;
    }
    & path:nth-child(3) {
      transform: rotate(-45deg) translateX(-17px) translateY(-5px);
    }
  }
}

@media (max-width: 800px) {
  .menu-icon {
    display: block;
  }
}
