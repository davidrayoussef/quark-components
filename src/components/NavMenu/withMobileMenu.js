import React, { Fragment } from 'react';
import Icon from '../Icon/Icon';
import style from './withMobile.css';

export const withMobileMenu = (Component) => {
  return class NavMenuWithMobile extends React.Component {
    state = {
      isOpen: false,
      mobileStyles: {}
    };

    handleMenuIconClick = (e) => {
      this.setState({
        isOpen: !this.state.isOpen
      });
    };

    handleMobileLinkClick = () => {
      this.setState({
        isOpen: false
      });
    };

    handleResize = () => {
      if (window.innerWidth < 800) {
        this.setState({
          mobileStyles: {
            width: '100vw',
            height: '100vh',
            position: 'fixed'
          }
        })
      }
      else {
        this.setState({
          mobileStyles: {}
        })
      }
    };

    componentDidMount() {
      this.handleResize();
      window.addEventListener('resize', this.handleResize);
    }

    componentWillUnmount() {
      window.removeEventListener('resize', this.handleResize);
    }

    render() {
      const { isOpen } = this.state;

      return (
        <Fragment>
          <Icon
            className={`${style['menu-icon']} ${isOpen ? style.close : ''}`}
            value="menu"
            onClick={this.handleMenuIconClick}
          />
          <Component
            isOpen={isOpen}
            handleMobileLinkClick={this.handleMobileLinkClick}
            mobileStyles={this.state.mobileStyles}
            {...this.props}
          />
        </Fragment>
      );
    }
  }
};
