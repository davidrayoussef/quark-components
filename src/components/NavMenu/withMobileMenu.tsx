import React, { Fragment } from 'react';

import { Icon } from '../Icon';

import { NavMenuProps, NavMenuWithMobileState } from '../../shared';

import style from './withMobile.css';

export const withMobileMenu = (
  Component: React.ComponentClass<NavMenuProps>
): React.ComponentClass<NavMenuProps> => {
  return class NavMenuWithMobile extends React.Component<
    NavMenuProps,
    NavMenuWithMobileState
  > {
    state: Readonly<NavMenuWithMobileState> = {
      isOpen: false,
      mobileStyles: {}
    };

    handleMenuIconClick = (): void => {
      this.setState({
        isOpen: !this.state.isOpen
      });
    };

    handleMobileLinkClick = (): void => {
      this.setState({
        isOpen: false
      });
    };

    handleResize = (): void => {
      if (window.matchMedia('(max-width: 800px)').matches) {
        this.setState({
          mobileStyles: {
            width: '100vw',
            height: '100vh',
            position: 'fixed',
            overflowY: 'auto',
            paddingBottom: '100%'
          }
        });
      } else {
        this.setState({
          mobileStyles: {}
        });
      }
    };

    componentDidMount(): void {
      this.handleResize();
      window.addEventListener('resize', this.handleResize);
    }

    componentWillUnmount(): void {
      window.removeEventListener('resize', this.handleResize);
    }

    render(): JSX.Element {
      const { isOpen } = this.state;
      return (
        <Fragment>
          <Icon
            value="menu"
            className={`${style.menuIcon} ${isOpen ? style.close : ''}`}
            onClick={this.handleMenuIconClick}
          />
          <Component
            {...this.props}
            isOpen={isOpen}
            handleMobileLinkClick={this.handleMobileLinkClick}
            mobileStyles={this.state.mobileStyles}
          />
        </Fragment>
      );
    }
  };
};
