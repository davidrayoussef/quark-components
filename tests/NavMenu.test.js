import React from 'react';
import { shallow } from 'enzyme';
import NavMenu from '../src/components/NavMenu/NavMenu';

describe('<NavMenu />', () => {
  it('should render', () => {
    const wrapper = shallow(<NavMenu data={[{ name: 'menuItem1' }]}/>);
    expect(wrapper);
  });
});
