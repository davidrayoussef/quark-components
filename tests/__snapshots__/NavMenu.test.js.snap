// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<React.Fragment>
  <aside
    className="sidebar"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <nav
      className="nav"
    >
      <a
        href="/nave menu item 1"
        key="nave menu item 1"
        onClick={[Function]}
        style={
          Object {
            "color": "rgb(52, 88, 226)",
          }
        }
      >
        nave menu item 1
      </a>
      <a
        href="/nave menu item 2"
        key="nave menu item 2"
        onClick={[Function]}
        style={
          Object {
            "color": "rgb(52, 88, 226)",
          }
        }
      >
        nave menu item 2
      </a>
      <a
        href="/nave menu item 3"
        key="nave menu item 3"
        onClick={[Function]}
        style={
          Object {
            "color": "rgb(52, 88, 226)",
          }
        }
      >
        nave menu item 3
      </a>
    </nav>
    <style>
      
    nav &gt; a:hover {
      background: rgba(52, 88, 226, .1);
    }
  
    </style>
  </aside>
</React.Fragment>
`;
